tortoisebeacon/weakcoin/weak_coin.go

```diff
@@ -4,290 +4,325 @@ import (
+func (wc *WeakCoin) FinishEpoch() {
+	wc.mu.Lock()
+	defer wc.mu.Unlock()
+	wc.allowances = nil
+	wc.coins = map[types.RoundID]bool{}
+	wc.round = 0
+}
 
-	message := Message{
-		Epoch:        epoch,
-		Round:        round,
-		MinerID:      wc.minerID,
-		VRFSignature: signedProposal,
+// StartRound process any buffered messages for this round and broadcast our proposal.
+func (wc *WeakCoin) StartRound(ctx context.Context, round types.RoundID) error {
+	wc.mu.Lock()
+	wc.logger.With().Info("started round",
+		log.Uint32("epoch_id", uint32(wc.epoch)),
+		log.Uint64("round_id", uint64(round)))
+	wc.round = round
+	wc.smallestProposal = nil
+	epoch := wc.epoch
+	for i, msg := range wc.nextRoundBuffer {
+		if msg.Epoch != wc.epoch || msg.Round != wc.round {
+			continue
+		}
+		if err := wc.updateProposal(msg); err != nil {
+			wc.logger.With().Debug("received invalid proposal",
+				log.Err(err),
+				log.Uint32("epoch_id", uint32(msg.Epoch)),
+				log.Uint64("round_id", uint64(msg.Round)))
+		}
+		wc.nextRoundBuffer[i].Epoch = 0
+		wc.nextRoundBuffer[i].Round = 0
+		wc.nextRoundBuffer[i].Unit = 0
+		wc.nextRoundBuffer[i].Signature = nil
 	}
+	wc.nextRoundBuffer = wc.nextRoundBuffer[:0]
+	wc.mu.Unlock()
+	return wc.publishProposal(ctx, epoch, round)
+}
 
-	serializedMessage, err := types.InterfaceToBytes(message)
+func (wc *WeakCoin) updateProposal(message Message) error {
+	buf := wc.encodeProposal(message.Epoch, message.Round, message.Unit)
+	miner, err := wc.verifier.Extract(buf, message.Signature)
 	if err != nil {
-		return fmt.Errorf("serialize weak coin message: %w", err)
+		return fmt.Errorf("can't recover miner id from signature %x: %w", message.Signature, err)
 	}
 
-	if err := wc.net.Broadcast(ctx, GossipProtocol, serializedMessage); err != nil {
-		return fmt.Errorf("broadcast weak coin message: %w", err)
+	allowed, exists := wc.allowances[string(miner.Bytes())]
+	if !exists || allowed < message.Unit {
+		return fmt.Errorf("miner %x is not allowed to submit proposal for unit %d", miner, message.Unit)
 	}
 
-	wc.Log.With().Info("Published proposal",
-		log.Uint64("epoch_id", uint64(epoch)),
-		log.Uint64("round_id", uint64(round)),
-		log.String("proposal", types.BytesToHash(signedProposal).ShortString()))
-
+	wc.updateSmallest(message.Signature)
 	return nil
 }
 
-func (wc *weakCoin) proposalExceedsThreshold(proposal []byte) bool {
-	proposalInt := new(big.Int).SetBytes(proposal)
-
-	return proposalInt.Cmp(wc.threshold) == 1
-}
-
-func (wc *weakCoin) generateProposal(epoch types.EpochID, round types.RoundID) ([]byte, error) {
-	proposal := bytes.Buffer{}
-
-	proposal.WriteString(proposalPrefix)
-
-	epochBuf := make([]byte, 8)
-	binary.LittleEndian.PutUint64(epochBuf, uint64(epoch))
-	if _, err := proposal.Write(epochBuf); err != nil {
-		return nil, fmt.Errorf("write epoch to buffer: %w", err)
+func (wc *WeakCoin) prepareProposal(epoch types.EpochID, round types.RoundID) (broadcast, smallest []byte) {
```

Comment:
Do we need a unit test for this?

Reply:
this code is tested. i can add regression test to verify that encoding doesn't change